### PR TITLE
Add Unit Test for Neural Network

### PR DESCRIPTION
**Pull request by jijoongmoon**
- Neural Network Initialization
- set Data Buffer
- set Optimizer

**Self evaluation:**
1. Build test:	 [X]Passed [ ]Failed [ ]Skipped
2. Run test:	 [X]Passed [ ]Failed [ ]Skipped

Signed-off-by: jijoong.moon <jijoong.moon@samsung.com>

### CODE REVIEW

File: nntrainer/include/databuffer.h [hunk 1]
```diff
@@ -130,6 +154,23 @@ class DataBuffer {
                          std::vector<std::vector<std::vector<float>>> &out_label, unsigned int batch,
                          unsigned int width, unsigned int height, unsigned int c_num);
 
+  /**
+   * @brief     set train data file name
+   * @param[in] path file path
+   * @param[in] type data type : DATA_TRAINING, DATA_VAL, DATA_TESTING
```
Comment (wooksong) [inline nntrainer/include/databuffer.h:160]:
```suggestion
   * @param[in] type data type : DATA_TRAINING, DATA_VAL, DATA_TEST, DATA_LABEL
```

File: nntrainer/include/neuralnet.h [hunk 2]
```diff
@@ -86,21 +94,21 @@ class NeuralNetwork {
...
+   * @brief     Get Optimizer
+   * @retval    Optimizer
+   */
   Optimizer getOptimizer() { return opt; };
 
   /**
-    * @brief     Get Learing rate
-    * @retval    Learning rate
-    */
+   * @brief     Get Learing rate
```
Comment (wooksong) [inline nntrainer/include/neuralnet.h:109]:
```suggestion
   * @brief     Get Learning rate
```

File: nntrainer/src/databuffer.cpp [hunk 3]
```diff
@@ -415,4 +416,50 @@ bool DataBuffer::getDataFromBuffer(BufferType type, std::vector<std::vector<std:
...
+        labels.push_back(data);
+      }
+    }
+      if (labels.size() != class_num) {
+        ml_loge("Error: number of label is not equal to number of class : %lu vs. %d", labels.size(), class_num);
+        return ML_ERROR_INVALID_PARAMETER;
+      }
+      break;
+    default:
+      break;
```
Comment (zhoonit) [inline nntrainer/src/databuffer.cpp:450]: I think
1. It is better to return `ML_ERROR_INVALID_PARAMETER` in case of default since it is not handled in this code.
2. there should be `case DATA_UNKNOWN`  with `ml_loge` for better debugging

File: nntrainer/src/neuralnet.cpp [hunk 4]
```diff
@@ -143,15 +143,15 @@ unsigned int parseType(std::string ll, InputType t) {
    *            "he_normal"  : He Normal Initialization
    *            "he_uniform"  : He Uniform Initialization
    */
-  std::array<std::string, 150> weight_ini_string = {"lecun_normal", "lecun_uniform", "xavier_normal", "xavier_uniform",
-                                                    "he_normal",    "he_uniform",    "unknown"};
+  std::array<std::string, 7> weight_ini_string = {"lecun_normal", "lecun_uniform", "xavier_normal", "xavier_uniform",
```
Comment (zhoonit) [inline nntrainer/src/neuralnet.cpp:146]: I think `layer_string` , `weight_ini_string`... should have same case unless there is any reason.

currently,
 `layer_string`:  PascalCase,
 `weight_ini_string`: snake_case,
 `others`: camel_case || lowercase

File: nntrainer/src/optimizer.cpp [hunk 5]
```diff
@@ -23,10 +23,39 @@
...
+  if (t == OptType::unknown) {
+    ml_loge("Error: Optimizer is unknown");
+    return ML_ERROR_INVALID_PARAMETER;
+  }
+  return status;
+}
+
+int Optimizer::setOptParam(OptParam p) {
+  int status = ML_ERROR_NONE;
+  if (p.learning_rate <= 0) {
```
Comment (zhoonit) [inline nntrainer/src/optimizer.cpp:42]: It seems learning_rate with 0 is acceptable. https://keras.io/optimizers/

File: test/unittest/unittest_nntrainer_internal.cpp [hunk 6]
```diff
@@ -46,10 +119,178 @@ TEST(nntrainer_NeuralNetwork, setConfig_02_n) {
...
+  p.weight_decay.lambda = 0.001;
+  status = op.setOptParam(p);
+  EXPECT_EQ(status, ML_ERROR_INVALID_PARAMETER);
+}
+
 /**
  * @brief Main gtest
  */
-int main(int argc, char **argv) {
+int main(int argc, char** argv) {
```
Comment (wooksong) [inline test/unittest/unittest_nntrainer_internal.cpp:293]: Any reason?

Comment (jijoongmoon) [reply test/unittest/unittest_nntrainer_internal.cpp:293]: Cause we are using clang-format-4.0. It changes automatically.

File: nntrainer/src/neuralnet.cpp (hunk 4 above)
Comment (jijoongmoon) [reply nntrainer/src/neuralnet.cpp:146]: right. I will change them later. Thanks.

File: nntrainer/src/databuffer.cpp (hunk 3 above)
Comment (jijoongmoon) [reply nntrainer/src/databuffer.cpp:450]: fixed.

File: nntrainer/src/optimizer.cpp (hunk 5 above)
Comment (jijoongmoon) [reply nntrainer/src/optimizer.cpp:42]: Well, I don't know why the keras get the zero learning rate. May be they have their own reason. However, it is not going to converge for the nntrainer if learning rate sets zero. If we will find the reason to set the learning rate is zero, we could change it then.

File: nntrainer/include/neuralnet.h (hunk 2 above)
Comment (jijoongmoon) [reply nntrainer/include/neuralnet.h]: fixed. Thanks.

File: nntrainer/include/databuffer.h (hunk 1 above)
Comment (jijoongmoon) [reply nntrainer/include/databuffer.h:147]: fixed.

File: debian/rules [hunk 7]
```diff
@@ -32,6 +32,10 @@ override_dh_auto_build:
 	ninja -C build
 
 override_dh_auto_test:
+	tar xzf ./packaging/trainset.tar.gz -C build
+	tar xzf ./packaging/valset.tar.gz -C build
+	tar xzf ./packaging/testset.tar.gz -C build
+	cp ./packaging/label.dat build
```
Comment (myungjoo) [inline debian/rules:38]: To avoid license complication and bloated git repo, you can create a "dataset" git repo and package its files in .deb and .rpm and refer to it at build-time (BuildRequires).

File: nntrainer/include/databuffer.h [hunk 8]
```diff
@@ -42,6 +43,16 @@ namespace nntrainer {
  */
 typedef enum { BUF_TRAIN, BUF_VAL, BUF_TEST, BUFF_UNKNOWN } BufferType;
 
+/**
+ * @brief     Enumeration of data type
+ *            0. DATA_TRAIN ( Data for training )
+ *            1. DATA_VAL ( Data for validation )
+ *            2. DATA_TEST ( Data for test )
+ *            3. DATA_LABEL ( Data for test )
+ *            3. Unknown
```
Comment (dongju-chae) [inline nntrainer/include/databuffer.h:52]: 4. ?

File: nntrainer/src/databuffer.cpp [hunk 9]
```diff
@@ -415,4 +416,53 @@ bool DataBuffer::getDataFromBuffer(BufferType type, std::vector<std::vector<std:
...
+        return ML_ERROR_INVALID_PARAMETER;
+      }
+    } break;
+    case DATA_UNKNOWN:
+    default:
+      ml_loge("Error: Data Type is unknown");
+      return ML_ERROR_INVALID_PARAMETER;
+      break;
+  }
+  return status;
```
Comment (dongju-chae) [inline nntrainer/src/databuffer.cpp:455]: It seems that `status` is always `ML_ERROR_NONE`. You may delete this variable or set `status` inside the switch statement.

File: debian/rules (hunk 7 above)
Comment (myungjoo) [reply debian/rules:38]: Anyway, it's fine to proceed with this and move files to another git repo later.

Comment (jijoongmoon) [reply debian/rules:38]: Yes. this is generated by the nntrainer itself. so that there is no license issues I guess and these are only for the testing. I intentionally make them small enough, but I will move them another git repo if more data is required.

File: nntrainer/src/databuffer.cpp (hunk 9 above)
Comment (jijoongmoon) [reply nntrainer/src/databuffer.cpp:455]: If it is then, I have to every case statement. so I prefer as it is.

File: nntrainer/include/databuffer.h (hunk 8 above)
Comment (jijoongmoon) [reply nntrainer/include/databuffer.h:52]: yes. But because there are PRs which related this, I will change it later. Thanks.
